pkg/network/common/egress_dns.go

```diff
@@ -117,43 +131,46 @@ func (e *EgressDNS) Sync() {
 				// Item needs to wait for this duration before it can be processed
 				duration = tm.Sub(now)
 			} else {
-				changed, err := e.Update(dnsName)
-				if err != nil {
-					utilruntime.HandleError(err)
-				}
-
-				if changed {
-					e.Updates <- updates
-				}
-				continue
+				e.dns.SetUpdating(dnsName)
+				go e.update(dnsName)
 			}
 		}
 
-		// Wait for the given duration or till something got added
+		// Wait for the the next query time, until there is a reply,
+		// or until a new name is added.
 		select {
+		case response := <-e.dnsResponse:
+			go e.handleDNSResponse(response)
 		case <-e.added:
+		case <-e.stopCh:
+			return
 		case <-time.After(duration):
 		}
 	}
 }
 
-func (e *EgressDNS) GetNextQueryTime() (time.Time, string, []EgressDNSUpdate, bool) {
-	e.lock.Lock()
-	defer e.lock.Unlock()
-	policyUpdates := make([]EgressDNSUpdate, 0)
-	tm, dnsName, timeSet := e.dns.GetNextQueryTime()
-	if !timeSet {
-		return tm, dnsName, nil, timeSet
+func (e *EgressDNS) handleDNSResponse(response DNSResponseNotification) {
+	trace := utiltrace.New(fmt.Sprintf("Handle DNS response notification for %q", response.Name))
+	defer trace.LogIfLong(dnsMapTraceThreshold)
+
+	if response.Changed {
+		updates := e.getEgressDNSUpdates(response.Name)
+		trace.Step("getEgressDNSUpdates")
+		e.Updates <- updates
 	}
 
+}
+
+func (e *EgressDNS) getEgressDNSUpdates(dnsName string) []EgressDNSUpdate {
+	policyUpdates := make([]EgressDNSUpdate, 0)
```

Comment:
This probably needs to grab a lock since it iterates over the dnsNamesToPolicies map which Add/Delete may muck around with?

Reply:
Good catch! Added the lock to it.